src/db.c

```diff
@@ -46,6 +48,76 @@ void updateLFU(robj *val) {
     val->lru = (LFUGetTimeInMinutes()<<8) | counter;
 }
 
+/* 
+ * Update histogram of keys-sizes
+ * 
+ * It is used to track the distribution of key sizes in the dataset. It is updated 
+ * every time key's length is modified. Available to user via INFO command. 
+ * 
+ * The histogram is a base-2 logarithmic histogram, with 64 bins. The i'th bin 
+ * represents the number of keys with a size in the range 2^i and 2^(i+1) 
+ * exclusive. oldLen/newLen must be smaller than 2^63. Each data type has its
+ * own histogram and it is per database (In addition, there is histogram per 
+ * slot for future use).
+ * 
+ * Examples to LEN values and corresponding bins in histogram: 
+ *               [1,2)->0 [2,4)->1 [4,8)->2 [8,16)->3
+ */
+void updateKeysizesHist(redisDb *db, int didx, uint32_t type, uint64_t oldLen, uint64_t newLen) {
+    serverAssert(unlikely(type < OBJ_TYPE_BASIC_MAX));
+
+    kvstoreDictMetadata *dictMeta = kvstoreGetDictMetadata(db->keys, didx);
+    kvstoreMetadata *kvstoreMeta = kvstoreGetMetadata(db->keys);
+
+    if (oldLen != 0) {
+        int old_bin = log2ceil(oldLen);
+        debugServerAssertWithInfo(server.current_client, NULL, old_bin < MAX_KEYSIZES_BINS);        
```

Comment:
maybe we shouldn't assume the size can't be larger than 2^48.
```suggestion
        int old_bin = log2ceil(oldLen);
        if (unlikely(old_bin >= MAX_KEYSIZES_BINS)) old_bin = MAX_KEYSIZES_BINS - 1;
```

Reply:
it will be much harder to catch such issue later on. I was even hold myself from being more aggressive with `serverAssert`. WDYT?

Reply:
i have no opinion with either way.

Reply:
Note also that i set MAX_KEYSIZES_BINS to be 48. i could have set it 64 and close my eyes. But i wanted that if we have a calculation issue, we better tackle it early on.

Reply:
i thought you just wanted to save some memory.

Reply:
also  :)